spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java

```diff
@@ -796,17 +801,24 @@ else if (replyTo instanceof TemporaryQueue || replyTo instanceof TemporaryTopic)
 		MessageConsumer messageConsumer = null;
 		try {
 			messageProducer = session.createProducer(requestDestination);
-			String correlationId = UUID.randomUUID().toString().replaceAll("'", "''");
 			Assert.state(this.correlationKey != null, "correlationKey must not be null");
 			String messageSelector = null;
-			if (this.correlationKey.equals("JMSCorrelationID")) {
-				jmsRequest.setJMSCorrelationID(correlationId);
-				messageSelector = "JMSCorrelationID = '" + correlationId + "'";
+			if (!this.correlationKey.equals("JMSCorrelationID*") || jmsRequest.getJMSCorrelationID() == null) {
+				String correlationId = UUID.randomUUID().toString().replaceAll("'", "''");
+				if (this.correlationKey.equals("JMSCorrelationID")) {
+					jmsRequest.setJMSCorrelationID(correlationId);
```

Comment:
Could it be applicable to check if there is a 'correlationId' property in `jmsRequest` and set it's value if not empty, otherwise - generated `correlationId`?